src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -893,6 +806,67 @@ FCIMPL4(Object*, RuntimeMethodHandle::InvokeMethod,
 }
 FCIMPLEND
 
+/// <summary>
+/// Convert a boxed value of {T} (which is either {T} or null) to a true boxed Nullable{T}.
+/// </summary>
+FCIMPL2(Object*, RuntimeMethodHandle::ReboxToNullable, Object* pBoxedValUNSAFE, ReflectClassBaseObject *pDestUNSAFE)
+{
+    FCALL_CONTRACT;
+
+    struct {
+        OBJECTREF pBoxed;
+        REFLECTCLASSBASEREF destType;
+        OBJECTREF retVal;
+    } gc;
+
+    gc.pBoxed = ObjectToOBJECTREF(pBoxedValUNSAFE);
+    gc.destType = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(pDestUNSAFE);
+    gc.retVal = NULL;
+
+    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+
+    MethodTable* destMT = gc.destType->GetType().AsMethodTable();
+
+    gc.retVal = destMT->Allocate();
+    void* buffer = gc.retVal->GetData();
+    BOOL result = Nullable::UnBox(buffer, gc.pBoxed, destMT);
+    _ASSERTE(result == TRUE);
+
+    HELPER_METHOD_FRAME_END();
+
+    return OBJECTREFToObject(gc.retVal);
+}
+FCIMPLEND
+
+/// <summary>
+/// For a true boxed Nullable{T}, re-box to a boxed {T} or null, otherwise just return the input.
+/// </summary>
+FCIMPL1(Object*, RuntimeMethodHandle::ReboxFromNullable, Object* pBoxedValUNSAFE)
+{
+    FCALL_CONTRACT;
+
+    struct {
+        OBJECTREF pBoxed;
+        OBJECTREF retVal;
+    } gc;
+
+    gc.pBoxed = ObjectToOBJECTREF(pBoxedValUNSAFE);
+    gc.retVal = gc.pBoxed;
+
+    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+
+    if (gc.pBoxed != NULL) {
+        MethodTable* retMT = gc.pBoxed->GetMethodTable();
+        if (Nullable::IsNullableType(retMT))
```

Comment:
It may be a good idea to test these two conditions outside the `HELPER_METHOD_FRAME` (better for performance). The actual reboxing is going to be rarely needed.

Reply:
Or even better, change `shouldCopyBack` to a bit field and call this Rebox helper only when the value is nullable that needs reboxing.

Reply:
At one point in the managed code I checked whether or not the value was a nullable (didn't track however) and the gain, if any, was basically not measurable even when I changed the code temporary to wrap it in an exaggerated loop- I would have needed to do some micro-benchmarks. Tracking whether the value is nullable may add a little overhead to the non-nullable cases, so I'd also need to do some micro-benchmarks.

Since the managed copyback is guarded by the per-parameter `bool` now at least we're not calling this method for the byvalue cases.

For now, I'll just move the conditions outside of `HELPER_METHOD_FRAME`. Thanks

Reply:
Update: a commit was pushed that avoids the call without degrading the perf of the non-nullable cases. The benchmark that takes 4 byref parameters is now ~10% faster.